akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -753,27 +756,39 @@ final class Partition[T](val outputPorts: Int, val partitioner: T ⇒ Int, val e
   override val shape: UniformFanOutShape[T, T] = UniformFanOutShape[T, T](in, out: _*)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+    val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
     private var outPendingElem: Any = null
     private var outPendingIdx: Int = _
     private var downstreamRunning = outputPorts
 
     def onPush() = {
       val elem = grab(in)
-      val idx = partitioner(elem)
-      if (idx < 0 || idx >= outputPorts) {
-        failStage(PartitionOutOfBoundsException(s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [${elem.getClass.getName}]."))
-      } else if (!isClosed(out(idx))) {
-        if (isAvailable(out(idx))) {
-          push(out(idx), elem)
-          if (out.exists(isAvailable(_)))
+      try {
+        val idx = partitioner(elem)
+        if (idx < 0 || idx >= outputPorts) {
+          failStage(PartitionOutOfBoundsException(s"partitioner must return an index in the range [0,${outputPorts - 1}]. returned: [$idx] for input [${elem.getClass.getName}]."))
+        } else if (!isClosed(out(idx))) {
+          if (isAvailable(out(idx))) {
+            push(out(idx), elem)
+            if (out.exists(isAvailable(_)))
+              pull(in)
+          } else {
+            outPendingElem = elem
+            outPendingIdx = idx
+          }
+
+        } else if (out.exists(isAvailable(_)))
+          pull(in)
+      } catch {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            failStage(ex)
+          case Supervision.Restart ⇒
+            pull(in)
```

Comment:
I think this is somewhat scary since it's not pulling in all normal cases. Perhaps we should only have the try-catch around the `partitioner` invocation?

Reply:
Yeah maybe good idea to move it just around that 👍